### ISSUE TITLE
[Discussion] Modify input manfiest yml schema and assemble workflow to support pre-installation of native plugins

### ISSUE THREAD

**Issue by joshpalis** [opensearch-project/opensearch-build#2849]
### Is your feature request related to a problem? Please describe

Now that the Job Scheduler project ownership has been transferred to the OpenSearch Core Extensibility Team, we and the community have made a decision to relocate Job Scheduler to native plugins [1]. This change will have an effect on how the full bundle of OpenSearch is assembled, since multiple plugins (ISM, AD, Reporting) have a dependency on Job Scheduler. 

Currently, the input build manifest includes Job Scheduler as a component[2], and this will be removed upon relocation. However, unless there is a mechanism to detect and pre-install native plugin dependencies, assembly of the aforementioned dependent plugins will fail. 

[1] https://github.com/opensearch-project/OpenSearch/issues/4218
[2] https://github.com/opensearch-project/opensearch-build/blob/main/manifests/2.4.0/opensearch-2.4.0.yml#L26

### Describe the solution you'd like

In order to support the assembly of the Job Scheduler dependent components, I propose modifications to the input manifest yml schema such that native plugins that other components depend on will also be included. Consequently, changes to the input manifest yml schema will necessitate changes to the bundle_OpenSearch assemble workflow, such that native plugins listed within the input manifest are pre-installed prior to the dependent components.

CC : @prudhvigodithi @peterzhuamazon Open to discussion on how best to handle these changes

**Comment by owaiskazi19** [comment]
Thanks for writing this up @joshpalis. How about including a flag named `depends_on` for the plugins such as ISM, AD which are dependent on Job Scheduler in the input manifest?
```yml
 - name: anomaly-detection
    repository: https://github.com/opensearch-project/anomaly-detection.git
    depends_on: ['Job-Scheduler']
    ref: '2.4'
    platforms:
      - linux
      - windows
    checks:
      - gradle:properties:version
      - gradle:dependencies:opensearch.version
```
While reading the input manifest we have to take into account the dependency and based on the flag we can either install or not the JS native plugin.

**Comment by prudhvigodithi** [comment]
Hey @owaiskazi19 something similar to what proposed [here](https://github.com/opensearch-project/opensearch-build/issues/441#issuecomment-1275192880) ?
Another way is to have new schema as following 
```
native_plugins:
- job-scheduler
- repository-s3
components:
  - name: OpenSearch
    repository: https://github.com/opensearch-project/OpenSearch.git
    ref: main
    checks:
      - gradle:publish
      - gradle:properties:version
  - name: common-utils
    repository: https://github.com/opensearch-project/common-utils.git
    ref: main
    platforms:
      - linux
      - windows
    checks:
      - gradle:publish
      - gradle:properties:version
```      
@dblock @bbarani

**Comment by owaiskazi19** [comment]
Hey @prudhvigodithi. Looks like we are on the same line with the scheme [here](https://github.com/opensearch-project/opensearch-build/issues/441#issuecomment-1275192880). 

For the above schema, I don't have any strong opinion(totally fine with it). Only just that it will create a little complex manifest file which can be simplified by using the flag `depends_on`. Let me know WDYT.

**Comment by joshpalis** [comment]
Thanks @owaiskazi19 and @prudhvigodithi for your contributions for this discussion. I am partial to Owais' proposal for the modifications to the input manifest yml schema. The modifications would be simplified and it places the onus of responsibility for components to define the plugins that they are dependent on, making it clear which components are dependent on which native plugin.

**Comment by prudhvigodithi** [comment]
[Triage]: Hey @joshpalis what is the targeted release with job-scheduler as native plugin, we should time this right to unblock 3.0.0 development.

**Comment by bbarani** [comment]
@joshpalis I would recommend to target 2.6.0 release since this is technically a breaking change for the build and development process so we need to synchronize the effort between multiple teams. CC: @CEHENKLE @peterzhuamazon @minalsha

**Comment by joshpalis** [comment]
Sure, I am fine with targeting 2.6.0 release. I have no strong opinions against this. @minalsha is this alright with you?

**Comment by minalsha** [comment]
@bbarani  we are working closely with folks across different teams. I don't see why we need to push it out. Lets target for 2.5.0 and see where we land on 01/02/2023. cc @CEHENKLE @prudhvigodithi @joshpalis @dagneyb

**Comment by bbarani** [comment]
@minalsha Is there a need to rush this change soon? This is not a new feature for users rather more of a change to existing plugin installation. This change needs to be synchronized across multiple versions. The available resources are currently working on 1.3.7 release along with Windows support hence I would not prioritize this for 2.5.0 release.

**Comment by saratvemulapalli** [comment]
> @saratvemulapalli Will this change be done only for 2.x and 3.x version. I assume 1.x versions are not affected by this change. Can you please confirm?

@bbarani you are right. This change will only go out for 2.x and above. This will not impact any 1.x releases. Job Scheduler will still be supported with the same maven co-ordinates for 1.x.

**Comment by joshpalis** [comment]
@bbarani @prudhvigodithi The relocation of Job Scheduler to native plugins has been determined to be a breaking change due to the necessary `build.gradle` modifications for dependent plugins. In our efforts to adhere to semantic versioning, we have modified our approach to relocate Job Scheduler to native plugins. Please refer to the updated section of [this issue](https://github.com/opensearch-project/OpenSearch/issues/5310) for additional information. 

Moving forward, modifications to the input manifest yml schema will still be necessary. Rather than the 2.6.0 release,  support for the pre-installation of native plugins shall only be needed for OpenSearch 3.x and onwards.

CC: @saratvemulapalli @minalsha

**Comment by gaiksaya** [comment]
[Proposal] Posting one of the approach here:

In order to add JS as a native plugin, a new schema needs to be introduced. Open search consist of a number of native plugins. In this case, we choose to install just one as a part of the distribution which is job scheduler (maybe more in future). Hence the manifest needs to document it.

One way is to have a `type` as one of the component key in the schema. The type key can be mandatory and the components can be defined accordingly. The following types seems valid:

* min (eg: OpenSearch, OpenSearch Dashboards)
* lib (eg: common-utils)
* core-plugin (eg: job scheduler, discovery-ec2)
* plugin (eg: anomaly detection, alerting)
* client: (eg: opensearch-high-level-rest client) → Example for future, out of scope for now.

Each type will define how the component is installed.

The functionality to install components that are of type core-plugin needs to be added to the build system.
* Building JS: From build perspective since native plugins are already built with the core we do not need to build again.
* Dependent plugins: For plugins with job scheduler as a dependency, it can retrieved from core dependencies which will already be taken care by plugins itself. The dependency will be retrieved from core modules so from build side if min is present we should be good.
* Assembling the distribution: In case of assembling the distribution, the JS needs to be installed from core plugins and then proceed with other plugin installations. 
* Publishing: For publishing and stuff, since it’s a part of native plugin now, it will continue to publish under https://repo1.maven.org/maven2/org/opensearch/plugin/opensearch-job-scheduler/ 

From offline discussion looks like the move will be happen from 2.x version but enforced from 3.x due to breaking changes.
We can decide if we want the new manifest schema should be introduced from 2.x version. Since build repository does not follow branching strategy to build (maybe it should), for 2.x the `type` for job-scheduler can be a `plugin` and then starting 3.0 it can be of type `core-plugin` 

The new input manifest can look like below:
```yaml
---
schema-version: '1.1'
build:
  name: OpenSearch
  version: 3.0.0
ci:
  image:
    name: opensearchstaging/ci-runner:ci-runner-centos7-opensearch-build-v2
    args: -e JAVA_HOME=/opt/java/openjdk-17
components:
  - name: OpenSearch
    repository: https://github.com/opensearch-project/OpenSearch.git
    ref: main
    type: min
    checks:
      - gradle:publish
      - gradle:properties:version
  - name: common-utils
    repository: https://github.com/opensearch-project/common-utils.git
    type: lib
    ref: main
    platforms:
      - linux
      - windows
    checks:
      - gradle:publish
      - gradle:properties:version
  - name: job-scheduler
    repository: https://github.com/opensearch-project/job-scheduler.git
    type: core-plugin
    ref: main
    platforms:
      - linux
      - windows
    checks:
      - gradle:properties:version
      - gradle:dependencies:opensearch.version
  - name: ml-commons
    repository: https://github.com/opensearch-project/ml-commons.git
    type: plugin
    ref: main
    platforms:
      - linux
      - windows
    checks:
      - gradle:properties:version
      - gradle:dependencies:opensearch.version: opensearch-ml-plugin
   
```

@saratvemulapalli @joshpalis @owaiskazi19  @dblock @opensearch-project/engineering-effectiveness  let me know what do you think about this. More approaches and recommendations are welcomed.
Thanks!

**Comment by dblock** [comment]
Generally adding a type is a clutch, I don't recommend it.

1. Does the current system work without changes and a working directory option, even if it means re-building the native plugin twice?
2. Since the native plugin is built, can it just be added to published artifacts instead as part of the build? What's missing if along with opensearch-min you get a job scheduler JAR/ZIP? Are verifications the only thing missing?

**Comment by bbarani** [comment]
@dblock Are you suggesting to integrate JS native plugin installation process outside of distribution build process? In that case, we will always assume that -min contains the job-scheduler (JS) pre-installed and install other plugins?

**Comment by saratvemulapalli** [comment]
This manifest is used for build and eventually assemble the distribution. 
Native plugin is a plugin, I dont think we'd want to differentiate between these two for builds. 

It changes with installation, the way they are installed. As long as we have right maven co-ordinates to install JS (as a native plugin), wouldn't that solve the problem? @gaiksaya 

The proposal also talks about order of execution which is a problem by itself.

**Comment by prudhvigodithi** [comment]
Hey all just circling back to this issue, thanks for multiple proposals added, @joshpalis just to refresh can you please add the timeline? Like what would be the next step for 2.6? the best chosen proposal has to be pushed before 2.6? As far as I know now the JS plugin is being tested as both as native and as a regular plugin, hence no change needed from build side. Will this strategy be same for 2.6 as well?
Thank you

**Comment by prudhvigodithi** [comment]
Hey @joshpalis just following up, can you please add your thoughts based on my [previous message](https://github.com/opensearch-project/opensearch-build/issues/2849#issuecomment-1468396997)?
Thank you

**Comment by dblock** [comment]
Do we need to change the manifest at all? Would it be simpler to install whichever native plugins we want as part of `install.sh` for OpenSearch core?

**Comment by prudhvigodithi** [comment]
Hey @dblock with the existing flow the [install.sh](https://github.com/opensearch-project/opensearch-build/blob/main/scripts/components/OpenSearch/install.sh) is called during [install_plugin](https://github.com/opensearch-project/opensearch-build/blob/main/src/assemble_workflow/bundle.py#L87) that has the logic just to run some `cp` commands, but the actual plugin installation for opensearch is happening from `install_plugin` method using [bundle_opensearch.py ](https://github.com/opensearch-project/opensearch-build/blob/main/src/assemble_workflow/bundle_opensearch.py#L24), this internally uses `opensearch-plugin` cli part of a random `tmp` folder (example: `/tmp/tmpvfijvu7o/opensearch-2.8.0/bin/opensearch-plugin`).

So looks to me like `install_plugin` from [bundle_opensearch.py](https://github.com/opensearch-project/opensearch-build/blob/main/src/assemble_workflow/bundle_opensearch.py#L24) has to be modified with native plugin list coming from a place (could be manifest) and works good.

Tested with `repository-s3` :
During assemble 
```
2023-06-07 18:47:07 INFO     Installed plugins: ['repository-s3', 'opensearch-job-scheduler']
```

Inside the final `tar` file.
```
opensearch-2.8.0/plugins/repository-s3/
opensearch-2.8.0/plugins/repository-s3/LICENSE.txt
opensearch-2.8.0/plugins/repository-s3/NOTICE.txt
opensearch-2.8.0/plugins/repository-s3/aws-java-sdk-core-1.12.270.jar
opensearch-2.8.0/plugins/repository-s3/aws-java-sdk-s3-1.12.270.jar
opensearch-2.8.0/plugins/repository-s3/aws-java-sdk-sts-1.12.270.jar
opensearch-2.8.0/plugins/repository-s3/commons-codec-1.15.jar
opensearch-2.8.0/plugins/repository-s3/commons-logging-1.2.jar
opensearch-2.8.0/plugins/repository-s3/httpclient-4.5.13.jar
opensearch-2.8.0/plugins/repository-s3/httpcore-4.4.15.jar
opensearch-2.8.0/plugins/repository-s3/jackson-annotations-2.15.1.jar
opensearch-2.8.0/plugins/repository-s3/jackson-databind-2.15.1.jar
opensearch-2.8.0/plugins/repository-s3/jaxb-api-2.3.1.jar
opensearch-2.8.0/plugins/repository-s3/jmespath-java-1.12.270.jar
opensearch-2.8.0/plugins/repository-s3/log4j-1.2-api-2.17.1.jar
opensearch-2.8.0/plugins/repository-s3/plugin-descriptor.properties
opensearch-2.8.0/plugins/repository-s3/plugin-security.policy
opensearch-2.8.0/plugins/repository-s3/repository-s3-2.8.0.jar
```

I'm open for any other solutions.
@gaiksaya @peterzhuamazon @zelinh please add your thoughts.

Thank you

**Comment by prudhvigodithi** [comment]
Another approach (first approach in previous [comment](https://github.com/opensearch-project/opensearch-build/issues/2849#issuecomment-1581647805)) we can go with is having a gradle task from OpenSearch core repo that can handle the desired native plugin installation code, and the script [install.sh](https://github.com/opensearch-project/opensearch-build/blob/main/scripts/components/OpenSearch/install.sh) can have this gradle task called which ensures the native plugins are installed.
Having this approach the ordering and requirement of right set of native plugins can be directly controlled from OpenSearch core repo.
@gaiksaya @dblock @joshpalis @saratvemulapalli

**Comment by peterzhuamazon** [comment]
> Hey @dblock with the existing flow the [install.sh](https://github.com/opensearch-project/opensearch-build/blob/main/scripts/components/OpenSearch/install.sh) is called during [install_plugin](https://github.com/opensearch-project/opensearch-build/blob/main/src/assemble_workflow/bundle.py#L87) that has the logic just to run some `cp` commands, but the actual plugin installation for opensearch is happening from `install_plugin` method using [bundle_opensearch.py ](https://github.com/opensearch-project/opensearch-build/blob/main/src/assemble_workflow/bundle_opensearch.py#L24), this internally uses `opensearch-plugin` cli part of a random `tmp` folder (example: `/tmp/tmpvfijvu7o/opensearch-2.8.0/bin/opensearch-plugin`).
> 
> So looks to me like `install_plugin` from [bundle_opensearch.py](https://github.com/opensearch-project/opensearch-build/blob/main/src/assemble_workflow/bundle_opensearch.py#L24) has to be modified with native plugin list coming from a place (could be manifest) and works good.
> 
> Tested with `repository-s3` : During assemble
> 
> ```
> 2023-06-07 18:47:07 INFO     Installed plugins: ['repository-s3', 'opensearch-job-scheduler']
> ```
> 
> Inside the final `tar` file.
> 
> ```
> opensearch-2.8.0/plugins/repository-s3/
> opensearch-2.8.0/plugins/repository-s3/LICENSE.txt
> opensearch-2.8.0/plugins/repository-s3/NOTICE.txt
> opensearch-2.8.0/plugins/repository-s3/aws-java-sdk-core-1.12.270.jar
> opensearch-2.8.0/plugins/repository-s3/aws-java-sdk-s3-1.12.270.jar
> opensearch-2.8.0/plugins/repository-s3/aws-java-sdk-sts-1.12.270.jar
> opensearch-2.8.0/plugins/repository-s3/commons-codec-1.15.jar
> opensearch-2.8.0/plugins/repository-s3/commons-logging-1.2.jar
> opensearch-2.8.0/plugins/repository-s3/httpclient-4.5.13.jar
> opensearch-2.8.0/plugins/repository-s3/httpcore-4.4.15.jar
> opensearch-2.8.0/plugins/repository-s3/jackson-annotations-2.15.1.jar
> opensearch-2.8.0/plugins/repository-s3/jackson-databind-2.15.1.jar
> opensearch-2.8.0/plugins/repository-s3/jaxb-api-2.3.1.jar
> opensearch-2.8.0/plugins/repository-s3/jmespath-java-1.12.270.jar
> opensearch-2.8.0/plugins/repository-s3/log4j-1.2-api-2.17.1.jar
> opensearch-2.8.0/plugins/repository-s3/plugin-descriptor.properties
> opensearch-2.8.0/plugins/repository-s3/plugin-security.policy
> opensearch-2.8.0/plugins/repository-s3/repository-s3-2.8.0.jar
> ```
> 
> I'm open for any other solutions. @gaiksaya @peterzhuamazon @zelinh please add your thoughts.
> 
> Thank you

I am ok with this approach, tho the input manifest needs changes to know the type of the plugins, per @gaiksaya suggestions.

**Comment by prudhvigodithi** [comment]
Hey @joshpalis coming form your comment https://github.com/opensearch-project/OpenSearch/issues/5310#issuecomment-1597518680, is the JS native plugin migration targeted to 3.0.0 release?

**Comment by joshpalis** [comment]
> Another approach (first approach in previous https://github.com/opensearch-project/opensearch-build/issues/2849#issuecomment-1581647805) we can go with is having a gradle task from OpenSearch core repo that can handle the desired native plugin installation code, and the script [install.sh](https://github.com/opensearch-project/opensearch-build/blob/main/scripts/components/OpenSearch/install.sh) can have this gradle task called which ensures the native plugins are installed.
Having this approach the ordering and requirement of right set of native plugins can be directly controlled from OpenSearch core repo.
@gaiksaya @dblock @joshpalis @saratvemulapalli

I support this approach, as it negates the need to modify the input manifest to add a type to differentiate plugins and native plugins. As @saratvemulapalli stated, they should be treated the same.

**Comment by prudhvigodithi** [comment]
Thanks Josh, in that case can you please coordinate from the core side to on board a task that can install the native plugins and then we can come back to the build side and update the [install.sh](https://github.com/opensearch-project/opensearch-build/blob/main/scripts/components/OpenSearch/install.sh) file with that gradle task ? 
Thank you

**Comment by bbarani** [comment]
@prudhvigodithi will look in this issue and see if it can be integrated to Gradle build for 3.x version.

**Comment by prudhvigodithi** [comment]
This issue remains on hold as the [proposal](https://github.com/opensearch-project/OpenSearch/issues/5310) and implementation for migrating job-scheduler as native plugin is still under review, also this migration is a breaking change and targeted to 3.0.0 release which is moved to Feb 18 2025 based on the release schedule https://github.com/opensearch-project/.github/issues/186. 

Its also worth exploring using of core `PersistentTaskPlugin` with adding scheduler capabilities and allow plugins use the core feature instead of job-scheduler.  Adding @peterzhuamazon  https://github.com/opensearch-project/job-scheduler/issues/147#issuecomment-1919192347.

Thanks
